### ISSUE TITLE
multiple GPU

### ISSUE THREAD

**Issue by rafariva** [jiny2001/dcscn-super-resolution#19]
Hi
I have a server with 4 GPU, trying to train. but i get an error (sorry the many captures of the error)

![image](https://user-images.githubusercontent.com/26292862/38829811-e1486f02-417f-11e8-8272-ef418fa226cb.png)
![image](https://user-images.githubusercontent.com/26292862/38829836-f26965ca-417f-11e8-99de-26476eaded4e.png)
![image](https://user-images.githubusercontent.com/26292862/38829845-fbcd3862-417f-11e8-9cf6-7c1e38c290c0.png)

![image](https://user-images.githubusercontent.com/26292862/38829778-c918bb58-417f-11e8-8306-9c848ee1847b.png)

![image](https://user-images.githubusercontent.com/26292862/38829680-84096256-417f-11e8-9a63-ce51900ab203.png)

how i can use the four? or why is this error ( i am giving a 5 images for test, and 20 for train, the images are 640x480)

**Comment by jiny2001** [comment]
HI,
Sorry, I don't know how to use multiple GPUs correctly. However, I'm pretty familiar to the error message. It's simply the batch is too big to your GPU memory.

That is kind of my fault. For tests during the training, I didn't use batch processing for feeding images. I mean, I put whole 5 images into the test batch at the same time. That's a memory consuming structure.

So here is the option.

Option 1. 
If you use ver1 branch, use "test_dataset" as default, "set5". When training finished, use evaluate
.py to check the performance. Run "evaluate.py --test_dataset [your images directory under "data"] [--other options]"

Option 2.
If you use ver2(currently master) branch, try "evaluate_dataset" arg instead of "test_dataset" arg for evaluating performance.
 
Let me know if you have any further questions. :)

**Comment by rafariva** [comment]
hi, looks like tensorflow-gpu automatically use all gpus (i still doing tests to confirm this). 
Already download your last version of the code. 
dont really understand what i should do for train with my new dataset. (do i have to reduce the batch size in args.py?)
i have a server with 96GB RAM, 4 GPU Nvidia tesla K20m

**Comment by jiny2001** [comment]
Hi,

OK, I updated code again, because it looks also nice to me. Could you try with your test images directory for --test_dataset ?

before: I split test images into batches (which consumes twice as large as original image) and load every images and feed to the model at once. 

after: I feed one image by one to the model.

In this version, test phase gets slower and logged loss value on tensorboard is not the whole loss. But you can put usual size of images as test data.

**Comment by rafariva** [comment]
hi

i am trying the traing with your last modification...

at the beginning, why it show that model? it even doesnt exists.
![image](https://user-images.githubusercontent.com/26292862/38998351-01842a0e-43b5-11e8-8598-fd1dd0362d5d.png)


looks like tensorflow-gpu detects my 4 nvidia
![image](https://user-images.githubusercontent.com/26292862/38998592-810fb702-43b5-11e8-820a-27fff6d3da23.png)

but even so, the training started with a warning about "low memory on GPU", but keep training .


now, in batch_data folder, doesnt create nothing??

**Comment by jiny2001** [comment]
Hi, the first message shows the name of your model. Your model will be saved during the training under the name.

After the training, you can compare the performance with the other model which was trained with the other parameter.

That is nice you could begin training.
Since 4G is not a satisfied size of memory, you will have the warning. When I use with 1060 with 3G RAM, some models couldn't even run on the card. Now I'm using 1080 Ti 11GB and I found this is the best card for personal use.

**Comment by jiny2001** [comment]
BTW, I updated the code last night since I found a bug in evaluation script.
And also I modified naming rule of model filename so now your model may not be compatible for latest codes.

But in that case, I can give you how to rename the model for latest code. Please let me know. thx.

**Comment by rafariva** [comment]
thanks for all you help. 

1) you said that the first message shows the name of my model and my  model will be saved during the training under the name. but before it saved its name with the name of my test_dataset

2) now in batch_data folder nothing is saving, its not necesary anymore?

**Comment by jiny2001** [comment]
1. I don't know.
GitHub repository contains some already trained model files. Maybe this is the one you were seeing?
Otherwise, it will be generated during the training.

2. No. it's not necessary for now.
However, pre-processing for building patches is really useful for training ( not for testing ). It will make the training significantly faster.

So maybe tomorrow or the day after tomorrow, I will make it available as an option. Will add to readme when I implemented.

**Comment by jiny2001** [comment]
Hi
Now I implemented "--build_batch True" option. It will work for medium size (640-1200) training data.
For large size (1200-2000) color images, it's better use convert_y.py to pre-process training dataset. 

I will close it. If you have any questions, feel free to ask. thx!

**Comment by rafariva** [comment]
Hello there,
i found out a way to asign the use of multiGPU. on your file tf_graph.py you must specificate the config, it could be like:

```
def init_session(self):
		config = tf.ConfigProto()
		config.gpu_options.allow_growth = True    ## just for use the necesary memory of gpu
		config.gpu_options.visible_device_list = '0,1,2,3'  ## this values depends of numbers of GPUs. ex if you have 4 GPU, and want to use 1,2 GPU, then just put = '1,2'.
		print("Session and graph initialized.")
		self.sess = tf.InteractiveSession(config=config, graph=tf.Graph())
```

maybe you can parametrizate in your args.py

**Comment by jiny2001** [comment]
Hi, thank you!

So, if I have 2 GPUs and I set it like "0,1", what will happen? Are some computation graphs assigned to each GPU automatically?

**Comment by rafariva** [comment]
If you hace 2 gpu. It would be same to leave it like already had. Would be the same uf you assign 0,1 (but you are right... If you have 2,u will have to put 0,1).
No quite shure if it assign each automaticcally. But, you can run your script assign Just gpu 0 and Also in other instance, on gpu 1

**Comment by jiny2001** [comment]
Ah, I understood. Thx.

Will implement it with command line arg. Since I have a PC with multiple GPU installed, I will try it.
Thank you so much. I'll keep open this ticket until I implement it.

**Comment by jiny2001** [comment]
Sorry for rate, I finally updated it. Thx!

**Comment by rafariva** [comment]
Hi, i see in your args.py the param init_session is int, but i think it should be string. how i should sent ex 0,3,4 GPU id?, if i send this values, you cast into int 
(maybe i am wrong)

**Comment by jiny2001** [comment]
I believe we don't need to send a list like that. Since my model only works with one GPU, there is no meaning if I set it as a list and set multiple ids. ( in that case maybe the most youngest id will be using?)

So I decided to implement to specify only one ID directly which will be used actually.

**Comment by rafariva** [comment]
yeah, but the idea should be, that some maybe has 4 GPU, and that person one to run the training using GPU 0 and 4, and for other training use GPU 3, and other training using the last GPU. 
or maybe someone has 8GPU. not always gona be the first one.
just saying.
(and actually i am using a share server that has 4 GPU, i had to modify your code to use just id=2,3, other guy is using the other two..

**Comment by jiny2001** [comment]
Hi, 
So, that was my original question. Does tensorflow use multiple GPU automatically?

For ex, I have 2 GPUs and when I put id="1,2" , does it calculated with multiple GPUs? It looks like the answer is No. In my case, just id=1 GPU runs for for every graph and the calculation time was same.

So I believe there is no meaning to specify multiple GPU ids because only one GPU is used and in this case it's better to specify that GPU id.

If you want assign each graph to each GPU, you have to do like this.
https://www.tensorflow.org/guide/using_gpu#using_multiple_gpus

So in this case, you have to fork the code and change it to 2 GPUs capable?

**Comment by rafariva** [comment]
when i specify to use all my 4 gpu, the code use all my gpu and i cant run anything else (gpu memory had been taken). but when i specify to use 0,1 in some training, and in the same time other traning using 2,3, it runs. 
I think that link is another way to do it. 

![image](https://user-images.githubusercontent.com/26292862/43786786-1548f756-9a2f-11e8-9927-c9413c46f75d.png)

**Comment by jiny2001** [comment]
Does it accelerate computation time when you specify 4GPUs list?
When you use sess = tf.Session(config=tf.ConfigProto(log_device_placement=True)) , you can log which device is assigned to which graph. Can you check it when you have a time??

**Comment by rafariva** [comment]
ok i will give it a try and let yo know, i will try short traning using one gpu, and the the 4 Gpu..
